### DISCUSSION TITLE
Can't get dependency_overrides to work

### DISCUSSION THREAD

**Discussion by moshevayner** [fastapi/fastapi#6515] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

Project structure (with only relevant files):
```
├── app
│   ├── __init__.py
│   ├── main.py
│   ├── routers
│   │   ├── __init__.py
│   │   ├── foo_router.py
│   ├── tests
│   │   ├── __init__.py
│   │   └── routers
│   │       ├── __init__.py
│   │       └── test_foo_router.py
```

`routers.foo_router` (minimal, only with relevant parts):
```python
from typing import List, Optional
from fastapi import APIRouter, Depends
from frontegg.fastapi.secure_access import FronteggSecurity, User

router = APIRouter()

@router.get("/bar", response_model=Optional[List[schemas.Foo]])
def list_bar(user: User = Depends(FronteggSecurity(permissions=["foo"]))):
    logger.debug(f"Authenticated user: {user}")
    return []
```

`app.main` (minimal, only with relevant parts):
```python
from fastapi import FastAPI
from app.routers import foo_router

app = FastAPI()

app.include_router(foo_router.router, prefix="/foo", tags=["Foo"])
```

`tests.routers.test_foo_router`:
```python

from fastapi.testclient import TestClient
from frontegg.fastapi.secure_access import FronteggSecurity, User

from app.main import app

def override_frontegg():
    return User(
        sub="user1",
        tenantId="tenant1",
        type="userToken",
        access_token="abc123",
    )

app.dependency_overrides[FronteggSecurity] = override_frontegg

def test_list_bar(client: TestClient) -> None:
    r = client.get("/foo/bar")

    response = r.json()
    assert r.status_code == 200
    assert response == []
```

### Description

Hey there!
I'm trying to apply that approach on my api, but something doesn't seem to work properly.
I found several other issues here as well as a couple of StackOverflow questions on this matter, but the suggested solutions don't seem to work for me. Unless there's something I've overlooked.
 
I've attached the relevant project structure and related files in the `Example Code` section.

However, when I run my tests, I get a 401 response, which means that the override isn't working as expected. It actually makes the call to my external auth provider (FrontEgg) although it shouldn't, based on the override that I've set.

Am I missing something here?

Thanks!

### Operating System

macOS

### Operating System Details

macOS Monterey 12.2 running on Apple MacBook Pro 13" with M1 chip.

### FastAPI Version

0.73.0

### Python Version

Python 3.9.9

**Comment by Kludex** [comment, marked as answer]
You cannot override dependencies that receive parameters. Well, you cannot override any dependency (object) that will have a different key value in a dictionary if you recreate it.

What you can do is:
```python
x = FronteggSecurity(permissions=["foo"])
```
Then you use `x` as dependency e.g. `Depends(x)`, and you can override it as well with `app.dependency_overrides[x] = y`.

**Reply by jonfulk** [reply]
Does this work the same when using Security with scopes as in :

```python
# main.py

from somewhere import SomeClass
# other imports...

auth = SomeClass()
security = Security(auth.verify, scopes=["write:something"])

@app.post(/route/to/biz)
def some_path_operation_func(request_body, payload=security):
    # do stuff
```

The class containing the class to verify is in another module and is imported here. In the test, I'm using a fixture that does this:

```python
from main import security
# other imports...

@pytest.fixture
def client_authenticated() -> TestClient:
    def skip_auth():
        return

    app.dependency_overrides[security] = skip_auth
    return TestClient(app)
```
The test uses this fixture as the test client, but it's still getting into the actual "verify" function that is supposed to be overridden. I've also tried importing the actual class both from main.py where it's used and from the module where it lives, but same result in all cases.

**Reply by jonfulk** [reply]
any ideas ^ @moshevayner ?

**Reply by jonfulk** [reply]
update: fixed this by getting rid of the class containing the `verify` method and just call `verify` directly when instantiating Security

```python
# main.py

from somewhere import verify
# other imports...

security = Security(verify, scopes=["write:something"])

@app.post(/route/to/biz)
def some_path_operation_func(request_body, payload=security):
    # do stuff
```

```python
@pytest.fixture
def client_authenticated() -> TestClient:
    def mock_verify():
        return {}

    app.dependency_overrides[verify] = mock_verify
    return TestClient(app)
```

**Comment by moshevayner** [comment]
@Kludex that worked for me, thanks!
I guess that the best solution would be to still support overriding a dependency that takes parameters (kinda like what `unittest.patch` does) if possible, but if not- I can still work with that proposed solution, definitely got me unblocked.
Thank you!
